cudax/include/cuda/experimental/__memory_resource/any_resource.cuh

```diff
@@ -41,310 +41,825 @@
 #include <cuda/std/__concepts/concept_macros.h>
-#include <cuda/std/__new_>
-#include <cuda/std/__type_traits/is_constructible.h>
-#include <cuda/std/__type_traits/is_nothrow_constructible.h>
-#include <cuda/std/__type_traits/is_same.h>
-#include <cuda/std/__type_traits/remove_cvref.h>
-#include <cuda/std/__type_traits/type_set.h>
-#include <cuda/std/__utility/exchange.h>
 #include <cuda/std/__utility/forward.h>
-#include <cuda/std/__utility/in_place.h>
+#include <cuda/std/optional>
 
-namespace cuda::experimental
-{
-template <class _Ty, class _Uy = _CUDA_VSTD::remove_cvref_t<_Ty>>
-_CCCL_INLINE_VAR constexpr bool __is_basic_any_resource = false;
-
-//! @rst
-//! .. _cudax-memory-resource-basic-any-resource:
-//!
-//! Base class for a type erased owning wrapper around a memory resource
-//! ---------------------------------------------------------------------
-//!
-//! ``basic_any_resource`` abstracts the differences between a resource and an async resource away, allowing efficient
-//! interoperability between the two.
-//!
-//! @endrst
-template <_CUDA_VMR::_AllocType _Alloc_type, class... _Properties>
-class basic_any_resource
-    : public _CUDA_VMR::_Resource_base<_Alloc_type, _CUDA_VMR::_WrapperType::_Owning>
-    , private _CUDA_VMR::_Filtered_vtable<_Properties...>
-{
-private:
-  static_assert(_CUDA_VMR::__contains_execution_space_property<_Properties...>,
-                "The properties of cuda::experimental::basic_any_resource must contain at least one execution "
-                "space property!");
-
-  template <_CUDA_VMR::_AllocType, class...>
-  friend class basic_any_resource;
-
-  template <class...>
-  friend struct _CUDA_VMR::_Resource_vtable;
+#include <cuda/experimental/__utility/basic_any.cuh>
 
-  using __vtable = _CUDA_VMR::_Filtered_vtable<_Properties...>;
+_CCCL_PUSH_MACROS
+#undef interface
 
-  //! @brief Validates that a set of \c _OtherProperties... is a superset of \c _Properties... .
-  template <class... _OtherProperties>
-  static constexpr bool __properties_match =
-    _CUDA_VSTD::__type_set_contains_v<_CUDA_VSTD::__make_type_set<_OtherProperties...>, _Properties...>;
+namespace cuda::experimental
+{
+#ifndef _CCCL_DOXYGEN_INVOKED // Do not document this
 
-  //! @brief Validates that a passed in \c _Resource satisfies the \c resource or \c async_resource concept respectively
-  //! as well as all properties in \c _Properties... .
-  template <class _Resource>
-  static constexpr bool __valid_resource =
-    _Alloc_type == _CUDA_VMR::_AllocType::_Async
-      ? _CUDA_VMR::async_resource_with<_Resource, _Properties...>
-      : _CUDA_VMR::resource_with<_Resource, _Properties...>;
+template <class _Property>
+using __property_result_t _CCCL_NODEBUG_ALIAS = _CUDA_VSTD::__type_call1< //
+  _CUDA_VSTD::conditional_t<cuda::property_with_value<_Property>,
+                            _CUDA_VSTD::__type_quote1<__property_value_t>,
+                            _CUDA_VSTD::__type_always<void>>,
+  _Property>;
```

Comment:
I find this overly complicated compared to just
```suggestion
using __property_result_t _CCCL_NODEBUG_ALIAS = _CUDA_VSTD::conditional_t<cuda::property_with_value<_Property>,__property_value_t<_Property>, void>;
```

Reply:
when `property_with_value<_Property>` is `false`, `__property_value_t<_Property>` is ill-formed.